### DISCUSSION TITLE
MU 4.0: "keep measures on the same system" should be staff dependent

### DISCUSSION THREAD

**Discussion by aige02** [musescore/MuseScore#16264] [Ideas]
Layout elements like system break and page break are part dependent. No information needs to be passed from the score to the parts. Conversely, section breaks do need to be passed to all parts. The "keep measures on the same system" (KMOSS) element in not like the section or page breaks in that a KMOSS needs to be passed from the score to the parts and vice versa. Neither is the KMOSS like the section break, since the KMOSS may not need to be passed to all parts.

Currently, the KMOSS is associated with a measure and is passed to all parts. This can be a problem since not all parts need to know about the KMOSS and the presence of a KMOSS can cause formatting deficiencies (specifically, preventing a desired system break) in some of the parts. If the KMOSS were associated with one or more staves (in addition to a measure), then the KMOSS can be passed only to those parts which would be affected by the KMOSS. Note that the KMOSS might affect more than one staff at a particular measure and might have to be associated with multiple staves at an affected measure. (This would be done automatically by a multi-measure rests on more than one stave or manually by the user from the Palette.).

This replaces the [MU forum feature request](https://musescore.org/en/comment/1172598).

**Comment by ghost** [comment]
Seems like a no-brainer to me. The KMOSS simply shouldn't transfer to parts, just as system and page breaks don't.

**Reply by Jojo-Schmitz** [reply]
May or may not work. Won't in context with multimeasure repeats (esp. a 2-measure repeat), which is mainly why they got (re-)implemented.
So it is staff and part detendant, not system dependant, like system, section and page breaks

**Reply by ghost** [reply]
Well the KMOSS is on the chopping block anyway as it's an hacky solution to problems which have a much more elegant solution.

**Reply by aige02** [reply]
I disagree; I believe that KMOSS should neither be like system or page breaks nor be on the chopping block.  I do agree that the current implementation of KMOSS is a "hacky solution".  Though a KMOSS is not like a section break that should be copied to all parts, neither is it like a system or page break that should never be copied to parts.
To repeat my suggestion above, a better implementation would be to associate a KMOSS with a measure in one staff - like staff text instead of system text.  If that measure is copied to the part, then the KMOSS is copied to the part as part of that staff-measure.  I see two notable conditions:
1. Multiple staves have a KMOSS at the same measure - this should not be a problem.  When the MU formatter detects one KMOSS when formatting a score or a part, any others at that measure can be ignored since they are all equivalent.  Like repeated staff text, the KMOSS symbol would appear above all flagged staves at the flagged measure.  When the staves are copied from the score to parts, then the additional KMOSS may be significant.  It might the only KMOSS at that measure in one or more parts.
2. Staggered KMOSS (for example, staff 1 repeats measures 1 and 2 numerous times, while staff 2 repeats measures 2 and 3 numerous times) - MU would have to recognize this condition and either:
   * input would disallow the staggering, or
   * (I recommend) formatting would be able to break a KMOSS if not otherwise possible to fit the data.  This might be necessary even without staggering if one staff has enough notes to fill a line while a second staff has a KMOSS that effectively requires those notes to be kept in the same system with another measure.
